File: 0701-0800/0794-Valid Tic-Tac-Toe State/0794-Valid Tic-Tac-Toe State.cpp
class Solution {
public:
    bool validTicTacToe(vector<string>& board) {
        int countX = 0, countO = 0;
        vector<int> rows(3), cols(3);
        int diag = 0, antidiag = 0;
        bool winX = false, winO = false;
        for (int i = 0; i < 3; ++i) {
            for (int j = 0; j < 3; ++j) {
                if (board[i][j] == 'X') {
                    ++countX;
                    ++rows[i];
                    ++cols[j];
                    if (i == j) {
                        ++diag;
                    }
                    if (i + j == 2) {
                        ++antidiag;
                    }
                }
                else if (board[i][j] == 'O') {
                    ++countO;
                    --rows[i];
                    --cols[j];
                    if (i == j) {
                        --diag;
                    }
                    if (i + j == 2) {
                        --antidiag;
                    }
                }
            }
        }
        if (countO > countX || countX - countO > 1) {
            return false;
        }
        
        if (rows[0] == 3 || rows[1] == 3 || rows[2] == 3 || cols[0] == 3 || cols[0] == 3 || cols[1] == 3 || cols[2] == 3 || diag == 3 || antidiag == 3) {
            winX = true;
        }
        if (rows[0] == -3 || rows[1] == -3 || rows[2] == -3 || cols[0] == -3 || cols[0] == -3 || cols[1] == -3 || cols[2] == -3 || diag == -3 || antidiag == -3) {
            winO = true;
        }
        
        if (winX) {
            if (winO) return false;
            if (countX == countO) return false;
        }
        else if (winO && countX > countO) {
            return false;
        }
        return true;
    }
};

File: 0701-0800/0794-Valid Tic-Tac-Toe State/0794-Valid Tic-Tac-Toe State.py
class Solution:
    def validTicTacToe(self, board: List[str]) -> bool:
        first, second = 'XO'
        countX = sum(row.count(first) for row in board)
        countO = sum(row.count(second) for row in board)
        if countO > countX or countX - countO > 1:
            return False
        def win(player):
            for i in range(3):
                if all(board[i][j] == player for j in range(3)):
                    return True
                if all(board[j][i] == player for j in range(3)):
                    return True
            return (player == board[1][1] == board[0][0] == board[2][2] or
                    player == board[1][1] == board[0][2] == board[2][0]) 
            
        if win(first) and countX != countO + 1:
            return False
        if win(second) and countX != countO:
            return False
        return True

class Solution2:
    def validTicTacToe(self, board: List[str]) -> bool:
        countX = countO = diag = antidiag = 0
        rows = [0]*3
        cols = [0]*3
        winX = winO = False
        for i in range(3):
            for j in range(3):
                if board[i][j] == 'X':
                    countX += 1
                    rows[i] += 1
                    cols[j] += 1
                    if i == j:
                        diag += 1
                    if i + j == 2:
                        antidiag += 1
                elif board[i][j] == 'O':
                    countO += 1
                    rows[i] -= 1
                    cols[j] -= 1
                    if i == j:
                        diag -= 1
                    if i + j == 2:
                        antidiag -= 1
        if countO > countX or countX - countO > 1:
            return False
        
        if any(rows[i] == 3 for i in range(3)) or any(cols[j] == 3 for j in range(3)) or diag == 3 or antidiag == 3:
            winX = True
            
        if any(rows[i] == -3 for i in range(3)) or any(cols[j] == -3 for j in range(3)) or diag == -3 or antidiag == -3:
            winO = True
        
        if winX and countX != countO + 1:
            return False
        if winO and countX != countO:
            return False
        return True
